### ISSUE TITLE
Floating point exceptions in CPP launch of pp_ttW  (FPE in vxxxxx function in SIMD mode)

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#701]
Floating point exceptions in CPP launch of pp_ttW 

This is in my branch nobm
```
[avalassi@itscrd80 gcc11.2/cvmfs] /data/avalassi/GPU2023/madgraph4gpuX/epochX/cudacpp> HRDCOD=1 tlau/lauX.sh -CPP nobm_pp_ttW.mad/
Execute lauX.sh for process nobm_pp_ttW.mad/ in directory /data/avalassi/GPU2023/madgraph4gpuX/epochX/cudacpp
...
Generating 10000 events with run name run_01
survey  run_01 
INFO: compile directory 
compile Source Directory
Using random number seed offset = 21
INFO: Running Survey 
Creating Jobs
Working on SubProcesses
INFO:     P1_gu_ttxwpd  
INFO: Building madevent in madevent_interface.py with 'CPP' matrix elements 
INFO:     P1_gd_ttxwmu  
Note: The following floating-point exceptions are signalling: IEEE_INVALID_FLAG IEEE_DIVIDE_BY_ZERO IEEE_UNDERFLOW_FLAG IEEE_DENORMAL
Note: The following floating-point exceptions are signalling: IEEE_INVALID_FLAG IEEE_DIVIDE_BY_ZERO IEEE_UNDERFLOW_FLAG IEEE_DENORMAL
...
```

**Comment by hageboeck** [comment]
I investigated this a little for
`epochX/cudacpp/nobm_pp_ttW.mad/SubProcesses/P1_gd_ttxwmu`

The first floating-point exception is a devision by zero here:
```c++
      const fptype_v pzpt = pvec3 / ( pp * pt ) * sqh * hel;
```
Caused by `pt = {0, 0, 0, 0}`.

It's a different process, but here's the relevant code:
https://github.com/madgraph5/madgraph4gpu/blob/603ff90199e1e169e0b45a2e48d971e5ac46dac4/epochX/cudacpp/gg_ttgg.mad/src/HelAmps_sm.h#L496-L529

The fp exception is in line 516, caused by line 497, where
```
const fptype_sv pt = fpsqrt( ( pvec1 * pvec1 ) + ( pvec2 * pvec2 ) );
pvec1 = @0x7fffffbd98e0: {0, 0, 0, 0}
pvec2 = @0x7fffffbd9900: {0, 0, 0, 0}
```

**Comment by valassi** [comment]
Hi @hageboeck thanks a lot!

Ok this looks like "my" bug, something that I introduced. I am a bit surprised because I thought I had tested this extensively including for values where pt=0 (testxxx.cc), but it seems there is an issue.

Definitely this code is ugly, as it makes BOTH the calculation with pT=0 and that with pT>0 and then chooses a posteriori... with the tiny problem that if pt+0 then the calculation assuming pt>0 gives a FP exception
https://github.com/madgraph5/madgraph4gpu/blob/603ff90199e1e169e0b45a2e48d971e5ac46dac4/epochX/cudacpp/gg_ttgg.mad/src/HelAmps_sm.h#L515
```
#ifndef MGONGPU_CPPSIMD
      if( pt != 0. )
      {
        const fptype pzpt = pvec3 / ( pp * pt ) * sqh * hel;
        vc[3] = cxmake( -pvec1 * pzpt, -nsv * pvec2 / pt * sqh );
        vc[4] = cxmake( -pvec2 * pzpt, nsv * pvec1 / pt * sqh );
      }
      else
      {
        vc[3] = cxmake( -hel * sqh, 0. );
        // NB: Do not use "abs" for floats! It returns an integer with no build warning! Use std::abs!
        //vc[4] = cxmake( 0, nsv * ( pvec3 < 0. ? -std::abs( sqh ) : std::abs( sqh ) ) ); // AV why abs here?
        vc[4] = cxmake( 0., nsv * ( pvec3 < 0. ? -sqh : sqh ) ); // AV: removed an abs here
      }
#else
      // Branch A: pt != 0.
      const fptype_v pzpt = pvec3 / ( pp * pt ) * sqh * hel;
      const cxtype_v vcA_3 = cxmake( -pvec1 * pzpt, -(fptype)nsv * pvec2 / pt * sqh );
      const cxtype_v vcA_4 = cxmake( -pvec2 * pzpt, (fptype)nsv * pvec1 / pt * sqh );
      // Branch B: pt == 0.
      const cxtype vcB_3 = cxmake( -(fptype)hel * sqh, 0 );
      const cxtype_v vcB_4 = cxmake( 0, (fptype)nsv * fpternary( ( pvec3 < 0 ), -sqh, sqh ) ); // AV: removed an abs here
      // Choose between the results from branch A and branch B
      const bool_v mask = ( pt != 0. );
      vc[3] = cxternary( mask, vcA_3, vcB_3 );
      vc[4] = cxternary( mask, vcA_4, vcB_4 );
#endif
```

There are TWO (or three) things TODO
- first extens the testxxx test and understand why it did not pick this up
- second, fix the issue
- (third, check if there are other places in ixx/oxx/vxx functions that use this trick of masks and that may give issues)

To fix it, one way may be to simply use a loop over ieppV=0...neppV-1 and forget about SIMD. 

Otherwise, a truly SIMD solution would require having two branches both of which do not send exceptions. In this case it is probably impossible. Anyway, the performance penalty is probably small: the vxxx functions are caled ONCE even if you have 1000 Feynamn diagrams...

**Comment by valassi** [comment]
PS About the checking other masks... I guess that my idea whan I implemented this was that the two "branches" were doing different calculations completely, not that if statement was there as a protection against FP exceptions. To be checked...

**Comment by valassi** [comment]
I remembered a discussion with @hageboeck where he mentioned that fortran enables FPE traps.

I am doing some tests with FP traps enabled in C++. I do see errors in testxxx.

Debugging is complex, things are not straightforward. For instance I get some FPEs in ixxxxx in debug mode, but ixxxxx goes through if not in debug mode...

**Comment by valassi** [comment]
I am making progress on this in MR #723. There are very large changes needed, essentially all xxx hardcoded implementations need some fixes

**Comment by valassi** [comment]
I am still making progress on MR #723. This is very long and complex, but I think I am almost done. Amongst other things
- I have enabled FPE signals in testxxx tests. As suggested by @hageboeck, one problem is that linking with fortran automatically enables FPEs, so it makes sense that we test FPEs also in our c++- only tests.
- I have PROFOUNDLY changed the implementations of ixxxx, oxxxx, vxxxx for SIMD. The point is that, as Stephan noted, there were some elements of vectors that were 0, and dividing by those would trigger an FPE. Previously we were ignoring this as we have fast math.
- One very peculiar issue that I also solved in ixx/oxx/vxx is that I DISABLED AUTOVECTORIZATION on these three functions. My new implementation has explicit loops over neppV elements. But then with -O3 (so only in no-debug builds, nit in debug beuilds!) autovectorization was kicking in and going back to the buggy SIMD version that issues an FPE! I disabled this with GCC attributes (I was unable to find a relevant pragma). This must still be fixed for clang and icx, see #724.
- In the testxxx.cc, I mad a ew changes too. First, I now have two separate sets of 16 tests (maximum SIMD on float), one with mass=0 and one with mass>0. This is because I do not want to mix ixx/oxx/vxx on masses which are valid for some events but not all. Second, I made some hacks to always make sure that the other 6 specifialized functions which expect pz>0 for instance, will still have that valid on all SIMD events.

I still need to do many things including
- backport to codegen
- regenrate all processes and test them
- test the actual pp_ttW suggetsed by Francesco that triggered this all issue
- port to clang and icx

**Comment by valassi** [comment]
PS A couple more things that are possibly to do
- I tried to set up gcov for testxxx, but this was difficult because ixxx/oxxx/vxxx are template functions in a separate header, and I was unable to get coverage for them
- One should consider enabling FPEs in the rest of the code, eg in all check.exe/gcheck.exe (and correlate it there to the number of nans and fast math #252 but also things like Cadna), or also in the other tests. In runTest at some point I was getting FPEs outside the testxxx test and in the compare test instead (but this was before I disabled auto vectorization in ixxx etc).

**Comment by valassi** [comment]
I have closed this issue in merging #723. I will document it better in MR #723.

Just in summary:
- this #701 was about FOUR floating point exceptions in nobm_pp_ttW
- MR #723 fixes ONE of them, the IEEE_DIVIDE_BY_ZERO (essentially, using a fake denominator instead of 0 in SIMD vectors, and using the `volatile` keyword to force the optimizer to calculate exactly that)
- there are still THREE floating point exceptions in nobm_pp_ttW, which will be followed up in #733 and in branch nobm that is in WIP MR #706

All in all, unfortunately, the nobm_pp_ttW process that had been suggested by Francesco Giuli in ATLAS is still not usable :-(

cc @roiser @oliviermattelaer @hageboeck
